src/main/app/controller/courts/PostcodesController.ts

```diff
@@ -0,0 +1,220 @@
+
+  public async get(
+    req: AuthedRequest,
+    res: Response,
+    searchValue = '',
+    postcodes: string[] = null,
+    error = '',
+    areasOfLaw: string[] = null,
+    courtTypes: string[] = null,
+    updated = false): Promise<void> {
+    const slug: string = req.params.slug as string;
+
+    const errors: Error[] = [];
+    // If we have an error from validation when adding/removing or moving postcodes,
+    // append it
+    if (error) {
+      errors.push({text: error});
+    }
+
+    if (!areasOfLaw) {
+      await req.scope.cradle.api.getCourtAreasOfLaw(slug)
+        .then((value: AreaOfLaw[]) => areasOfLaw = value.map(ct => ct.name))
+        .catch(() => errors.push({text: this.getCourtAreasOfLawErrorMsg}));
+    }
+
+    if (areasOfLaw) {
+      areasOfLaw = this.checkCountyAreasOfLaw(areasOfLaw).map(ct => ct.replace(/\s/g, '_'));
+      if(!areasOfLaw.length){
+        errors.push({text: this.familyAreaOfLawErrorMsg});
+      }
+    }
+
+    if (!courtTypes) {
+      await req.scope.cradle.api.getCourtCourtTypes(slug)
+        .then((value: CourtType[]) => courtTypes = value.map(ct => ct.name.replace(/\s/g, '_')))
+        .catch(() => errors.push({text: this.getCourtTypesErrorMsg}));
+    }
+
+    if (!postcodes)
+      await req.scope.cradle.api.getPostcodes(slug)
+        .then((value: string[]) => postcodes = value)
+        .catch(() => errors.push({text: this.getPostcodesErrorMsg}));
+
+    const courts = await req.scope.cradle.api.getCourts();
+
+    const pageData: PostcodeData = {
+      postcodes: postcodes,
+      courts: courts,
+      slug: slug,
+      errors: errors,
+      updated: updated,
+      searchValue: searchValue,
+      isEnabled: !courtTypes.length ? false :courtTypes.some(ct => ct === 'County_Court'),
+      areasOfLaw: areasOfLaw,
+      courtTypes: courtTypes
+    };
+    res.render('courts/tabs/postcodesContent', pageData);
+  }
+
+  public async post(
+    req: AuthedRequest,
+    res: Response): Promise<void> {
+
+    const existingPostcodes: string[] = req.body.existingPostcodes?.split(',') ?? [];
+    const courtTypes = req.body.courtTypes?.split(',') ?? [];
+    const areasOfLaw = req.body.areasOfLaw?.split(',') ?? [];
+    if (!CSRF.verify(req.body.csrfToken)) {
+      return this.get(req, res, '', existingPostcodes, this.addErrorMsg, areasOfLaw, courtTypes);
+    }
+
+    // If there is no user input
+    const newPostcodes = req.body.newPostcodes ?? [];
+    if (newPostcodes.length == 0) {
+      return this.get(req, res, '', existingPostcodes, this.noPostcodeErrorMsg,
+        areasOfLaw, courtTypes, true);
+    }
+
+    // We should only send a postcode that is of at least size 7 (without spaces) min 2 characters
+    const newPostcodesArray = newPostcodes.replace(/\s/g, '').toUpperCase().split(',');
+    const invalidPostcodes = newPostcodesArray.filter((value: string) => value.length < 2 || value.length > 7);
+    if(invalidPostcodes.length > 0) {
```

Comment:
I'd be tempted to move validation to a separate method to reduce some of the bulk in this method.